transaction/memo.go

```diff
@@ -0,0 +1,134 @@
+
+var TraderAPIMemoProgram = solana.MustPublicKeyFromBase58("HQ2UUt18uJqKaQFJhgV9zaTdQxUZjNrsKFgoEDquBkcx")
+
+// CreateTraderAPIMemoInstruction generates a transaction instruction that places a memo in the transaction log
+// Having a memo instruction with signals Trader-API usage is required
+func CreateTraderAPIMemoInstruction(msg string) solana.Instruction {
+	if msg == "" {
+		msg = BxMemoMarkerMsg
+	}
+	buf := new(bytes.Buffer)
+	buf.Write([]byte(msg))
+
+	instruction := &solana.GenericInstruction{
+		AccountValues: nil,
+		ProgID:        TraderAPIMemoProgram,
+		DataBytes:     buf.Bytes(),
+	}
+
+	return instruction
+}
+
+func AddMemo(instructions []solana.Instruction, memoContent string, blockhash solana.Hash, owner solana.PublicKey, privateKeys map[solana.PublicKey]solana.PrivateKey) (string, error) {
+	memo := CreateTraderAPIMemoInstruction(memoContent)
+
+	instructions = append(instructions, memo)
+
+	txnBytes, err := buildFullySignedTxn(blockhash, owner, instructions, privateKeys)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(txnBytes), nil
+}
+
+// AddMemoToSerializedTxn adds memo instruction to a serialized transaction, it's primarily used if the user
+// doesn't want to interact with Trader-API directly
+func AddMemoToSerializedTxn(programIDIndex uint16, txBase64, memoContent string,
+	owner solana.PublicKey, privateKeys map[solana.PublicKey]solana.PrivateKey) (string, error) {
+	signedTxBytes, err := solanarpc.DataBytesOrJSONFromBase64(txBase64)
+	if err != nil {
+		return "", err
+	}
+	unsignedTx := solanarpc.TransactionWithMeta{Transaction: signedTxBytes}
+	solanaTx, err := unsignedTx.GetTransaction()
+	if err != nil {
+		return "", err
+	}
+	progID, err := solanaTx.Message.ResolveProgramIDIndex(programIDIndex)
+	if err != nil {
+		return "", err
+	}
+
+	memo := CreateTraderAPIMemoInstruction(memoContent)
+	var instructions []solana.Instruction
+
+	for _, cmpInst := range solanaTx.Message.Instructions {
+		accounts := cmpInst.ResolveInstructionAccounts(&solanaTx.Message)
+		instructions = append(instructions, &solana.GenericInstruction{
+			AccountValues: accounts,
+			ProgID:        progID,
+			DataBytes:     cmpInst.Data,
+		})
+	}
+	instructions = append(instructions, memo)
+	txnBytes, err := buildFullySignedTxn(solanaTx.Message.RecentBlockhash, owner, instructions, privateKeys)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(txnBytes), nil
+}
+
+func buildFullySignedTxn(recentBlockHash solana.Hash, owner solana.PublicKey, instructions []solana.Instruction, privateKeys map[solana.PublicKey]solana.PrivateKey) ([]byte, error) {
+
+	txBuilder := solana.NewTransactionBuilder()
+	for _, inst := range instructions {
+		txBuilder.AddInstruction(inst)
+	}
+
+	txBuilder.SetRecentBlockHash(recentBlockHash)
+	txBuilder.SetFeePayer(owner)
+
+	tx, err := txBuilder.Build()
+	if err != nil {
+		return nil, err
+	}
+	if privateKeys != nil {
```

Comment:
lines below this point can be simplified with `tx.Sign()` function

Reply:
bump, the `solana-go` library can simplify all this code. only reason the rest of this exist elsewhere in the codebase is because we needed to partially sign it but we can use the built-in function here instead

Reply:
can now delete this `if` condition